chain/client/src/stateless_validation/state_witness_tracker.rs

```diff
@@ -0,0 +1,213 @@
+pub struct ChunkStateWitnessTracker {
+    witnesses: LruCache<ChunkStateWitnessKey, ChunkStateWitnessRecord>,
+    clock: Clock,
+}
+
+impl ChunkStateWitnessTracker {
+    pub fn new(clock: Clock) -> Self {
+        Self {
+            witnesses: LruCache::new(CHUNK_STATE_WITNESS_MAX_RECORD_COUNT),
+            clock,
+        }
+    }
+
+    /// Adds a new witness message to track.
+    pub fn record_witness_sent(&mut self, witness: &ChunkStateWitness, witness_size_in_bytes: usize,
+                               num_validators: usize) -> () {
+        let key = ChunkStateWitnessKey::new(witness);
+        tracing::debug!(target: "state_witness_tracker", "Sent state witness: {:?} of {} bytes",
+            witness.inner.chunk_header.chunk_hash(), witness_size_in_bytes);
+        self.witnesses.put(key, ChunkStateWitnessRecord {
+            num_validators,
+            witness_size: witness_size_in_bytes,
+            sent_timestamp: self.clock.now(),
+        });
+    }
+
+    /// Handles an ack message for the witness. Calculates the round-trip duration and
+    /// records it in the corresponding metric.
+    pub fn on_witness_ack_received(&mut self, ack: ChunkStateWitnessAck) -> () {
+        let key = ChunkStateWitnessKey { chunk_hash: ack.chunk_hash };
+        tracing::debug!(target: "state_witness_tracker", "Received ack for state witness: {:?}",
+            key.chunk_hash);
+        if let Some(mut record) = self.witnesses.pop(&key) {
+            assert!(record.num_validators > 0);
+
+            Self::update_roundtrip_time_metric(&record, &self.clock);
+
+            // Cleanup the record if we received the acks from all the validators.
+            if let Some(remaining) = record.num_validators.checked_sub(1) {
+                if remaining > 0 {
+                    record.num_validators = remaining;
+                    self.witnesses.put(key.clone(), record);
+                }
+            }
+        }
+    }
+
+    /// Records the roundtrip time in metrics.
+    fn update_roundtrip_time_metric(record: &ChunkStateWitnessRecord, clock: &Clock) -> () {
+        let received_time = clock.now();
+        if received_time > record.sent_timestamp {
+            CHUNK_STATE_WITNESS_NETWORK_ROUNDTRIP_TIME
+                .with_label_values(&[witness_size_bucket(record.witness_size)])
+                .observe((received_time - record.sent_timestamp).as_seconds_f64());
+        }
+    }
+
+    #[cfg(test)]
+    fn get_record_for_witness(&mut self, witness: &ChunkStateWitness) -> Option<&ChunkStateWitnessRecord> {
+        let key = ChunkStateWitnessKey::new(witness);
+        return self.witnesses.get(&key);
+    }
+}
+
+
+/// Buckets for state-witness size.
+// TODO: Use size::Size to represent sizes.
+static SIZE_IN_BYTES_TO_BUCKET: &'static [(usize, &str)] = &[
+    (1_000, "<1KB"),
+    (10_000, "1-10KB"),
+    (100_000, "10-100KB"),
+    (1_000_000, "100KB-1MB"),
+    (5_000_000, "1-5MB"),
```

Comment:
most of "interesting" state witness cases based on current mainnet traffic will be from 1 to 3 MB, so I suggest splitting that range: "1-2MB", "2-3MB", "3-5MB"

Reply:
done (also added 3-4 and 4-5, should not be a big deal as long as #buckets is fixed)